source/simulator/assemblers/newton_stokes.cc

```diff
@@ -0,0 +1,677 @@
+{
+  namespace Assemblers
+  {
+    template <int dim>
+    void
+    NewtonStokesAssembler<dim>::
+    preconditioner (const double                                             pressure_scaling,
+                    internal::Assembly::Scratch::StokesPreconditioner<dim>  &scratch,
+                    internal::Assembly::CopyData::StokesPreconditioner<dim> &data,
+                    const Parameters<dim> &parameters) const
+    {
+      const Introspection<dim> &introspection = this->introspection();
+      const FiniteElement<dim> &fe = this->get_fe();
+      const unsigned int stokes_dofs_per_cell = data.local_dof_indices.size();
+      const unsigned int n_q_points           = scratch.finite_element_values.n_quadrature_points;
+      const double theta = parameters.newton_theta;
+
+      // First loop over all dofs and find those that are in the Stokes system
+      // save the component (pressure and dim velocities) each belongs to.
+      for (unsigned int i = 0, i_stokes = 0; i_stokes < stokes_dofs_per_cell; /*increment at end of loop*/)
+        {
+          if (introspection.is_stokes_component(fe.system_to_component_index(i).first))
+            {
+              scratch.dof_component_indices[i_stokes] = fe.system_to_component_index(i).first;
+              ++i_stokes;
+            }
+          ++i;
+        }
+
+      // Loop over all quadrature points and assemble their contributions to
+      // the preconditioner matrix
+      for (unsigned int q = 0; q < n_q_points; ++q)
+        {
+          for (unsigned int i = 0, i_stokes = 0; i_stokes < stokes_dofs_per_cell; /*increment at end of loop*/)
+            {
+              if (introspection.is_stokes_component(fe.system_to_component_index(i).first))
+                {
+                  scratch.grads_phi_u[i_stokes] =
+                    scratch.finite_element_values[introspection.extractors
+                                                  .velocities].symmetric_gradient(i, q);
+                  scratch.phi_p[i_stokes] = scratch.finite_element_values[introspection
+                                                                          .extractors.pressure].value(i, q);
+                  ++i_stokes;
+                }
+              ++i;
+            }
+
+          const double eta = scratch.material_model_outputs.viscosities[q];
+          const double one_over_eta = 1. / eta;
+
+          const SymmetricTensor<4, dim> &stress_strain_director = scratch
+                                                                  .material_model_outputs.stress_strain_directors[q];
+          const bool use_tensor = (stress_strain_director
+                                   != dealii::identity_tensor<dim>());
+
+          const double JxW = scratch.finite_element_values.JxW(q);
+
+          // TODO: Find out why in this version of ASPECT adding the derivative to the preconditioning
+          //is way worse than the normal preconitioning
+          if (true)//theta == 0)
```

Comment:
What happens to this comment? No need to solve this now if it does not work, but should we keep the `else` part out of the code for now?